engine/class_modules/sc_death_knight.cpp

```diff
@@ -4790,64 +4740,56 @@ struct unholy_pact_buff_t : public buff_t
     {
       damage -> execute();
     } );
+    // Unholy pact ticks twice on buff expiration, hence the following override
+    // https://github.com/SimCMinMax/WoW-BugTracker/issues/675
+    set_stack_change_callback( [ this, p ] ( buff_t*, int, int new_ ) 
+    {
+      if ( !new_ )
+      {
+        damage -> execute();
+        damage -> execute();
+      }
+    } );
   }
-
-  // Unholy pact ticks twice on buff expiration, hence the following override
-  // https://github.com/SimCMinMax/WoW-BugTracker/issues/675
-  void expire_override( int expiration_stacks, timespan_t remaining_duration ) override
-  {
-    damage -> schedule_execute();
-    damage -> schedule_execute();
-    buff_t::expire_override( expiration_stacks, remaining_duration );
-  }
+private:
+    propagate_const<action_t*> damage;
 };
 
-struct dark_transformation_damage_t : public death_knight_spell_t
+struct dark_transformation_damage_t final : public death_knight_spell_t
 {
   dark_transformation_damage_t( util::string_view name, death_knight_t* p ) :
-    death_knight_spell_t( name, p, p -> find_spell( 344955 ) )
+    death_knight_spell_t( name, p, p -> spell.dark_transformation_damage )
   {
     background = dual = true;
     aoe = as<int>( data().effectN( 2 ).base_value() );
   }
 };
 
 // Even though the buff is tied to the pet ingame, it's simpler to add it to the player
-struct dark_transformation_buff_t : public buff_t
+struct dark_transformation_buff_t final : public buff_t
 {
   dark_transformation_buff_t( death_knight_t* p ) :
     buff_t( p, "dark_transformation", p -> talent.unholy.dark_transformation )
   {
     set_default_value_from_effect( 1 );
     cooldown -> duration = 0_ms; // Handled by the player ability
-  }
-
-  // Unlike the player buff, Frenzied Monstrosity follows Dark Trasnformation uptime on the pet
-  // even with effects that increase the buff's duration
-  bool trigger( int s, double v, double c, timespan_t d ) override
-  {
-    death_knight_t* p = debug_cast<death_knight_t*>( player );
-
-    if ( p -> talent.unholy.ghoulish_frenzy.ok() )
-      debug_cast<pets::ghoul_pet_t*>( p -> pets.ghoul_pet ) -> ghoulish_frenzy -> trigger();
-
-    if ( p -> talent.unholy.ghoulish_frenzy.ok() )
-      p -> buffs.ghoulish_frenzy -> trigger();
-
-    return buff_t::trigger( s, v, c, d );
-  }
-
-  void expire_override( int, timespan_t ) override
-  {
-    death_knight_t* p = debug_cast<death_knight_t*>( player );
-
-    debug_cast<pets::ghoul_pet_t*>( p -> pets.ghoul_pet ) -> ghoulish_frenzy -> expire();
-
-    p -> buffs.ghoulish_frenzy -> expire();
+    set_stack_change_callback( [ this, p ] ( buff_t*, int, int new_ ) 
+    {
+      if ( new_ && p -> talent.unholy.ghoulish_frenzy.ok() )
+      {
+        p -> buffs.ghoulish_frenzy -> trigger();
+        debug_cast<pets::ghoul_pet_t*>( p -> pets.ghoul_pet ) -> ghoulish_frenzy -> trigger();
+      }
+      else
+      {
+        p -> buffs.ghoulish_frenzy -> expire();
```

Comment:
This block should be guarded by a check for the ghoulish frenzy talent so sims that don't have GF aren't making repeated cancel calls for a buff that doesn't get used.